### PR TITLE
Scatter Plot Color and Autosize Legend

### PR DESCRIPTION
**Pull request by MayaGans**
- [X] Add autosize to spec object to compute legend in plotObject div width
 - [x] Deal with default color issue #92 

I'm not sure if you'll like this solution but I:

Screen Shot 2020-07-31 at 8 54 03 AM

1. Changed the UI so the color default is `""`
2. Changed the generator to return valid color OR `" "`
3. Created validatorColor which looks for valid column OR `" "`
4. Changed TransformBlock, if `(color == " ")` a color isn't plotted

Now the user can toggle between adding a color and removing the color. **[Before once the user added a color, there was no way to then remove the color because re-writing "color" in the field was invalid!]**

Does this make sense? Happy to discuss alternative solutions!

### CODE REVIEW

**Comment by gvwilson** [comment]
I think this makes sense, and I can't think of a better approach - starting with an invalid color name like `…color…` would also work, but... I also need to do some work to figure out how to handle the case where there actually _is_ a column called `column` in a table :-/

**Comment by MayaGans** [comment]
The problem with starting with an invalid name is what if I want to plot penguins by species but then change my mind and I don't want to anymore. I can't type `...color...` to remove the color since it's invalid, and dragging an entirely new plot block seems.... annoying. Guess I'm liking my fix more and more.... 

PS I also like this UI where fields that start off blank are optional

File: blocks/blocks.js [hunk 1]
```diff
@@ -196,6 +195,20 @@ const createValidators = () => {
     }
   }
 
+  // Create a color validator
+  // This can either be a column or " "
+  const validateColor = (columnName, pattern) => {
```
Comment (gvwilson) [inline blocks/blocks.js:200]: `validateColor` requires two parameters, but is called with only one on line 224.

File: blocks/blocks.js [hunk 2]
```diff
@@ -196,6 +195,20 @@ const createValidators = () => {
...
+    return function () {
+      const field = this.getField(columnName)
+      field.setValidator((newValue) => {
+        if (newValue.match(pattern) || " ") {
```
Comment (gvwilson) [inline blocks/blocks.js:204]: This expression is always true: if `newValue.match(pattern)` fails, it then or's with a non-empty string, which is truthy.

File: blocks/blocks.js [hunk 3]
```diff
@@ -208,6 +221,7 @@ const createValidators = () => {
   Blockly.Extensions.register('validate_STDDEV', _createNonNeg('STDDEV'))
 
   Blockly.Extensions.register('validate_DATE', validateDate('DATE'))
+  Blockly.Extensions.register('validate_COLOR', validateColor('COLOR'))
```
Comment (gvwilson) [inline blocks/blocks.js:224]: `validateColor` requires a pattern but none is given.

File: test/test_transform.js [hunk 4]
```diff
@@ -253,7 +253,7 @@ describe('build plots', () => {
                  `Wrong X axis`)
     assert.equal(plot.layer[0].encoding.y.field, 'green',
                  `Wrong Y axis`)
-    assert(!('color' in plot.layer[0].encoding),
+    assert(!(' ' in plot.layer[0].encoding),
```
Comment (gvwilson) [inline test/test_transform.js:256]: hm - this is going to be hard for the next reader to figure out...

Comment (MayaGans) [reply test/test_transform.js]: does `assert(plot.layer[0].encoding.color.field === null, `Should not have color`)` resolve this/make it more human read-able? [see latest commits]

File: blocks/blocks.js (hunk 3 above)
Comment (MayaGans) [reply blocks/blocks.js]: I'm using the same pattern now as the single column but also allow for an empty string in new code - see latest commits. Thanks for the read through!

File: test/test_transform.js (hunk 4 above)
Comment (gvwilson) [reply test/test_transform.js]: You can use `assert.strictEqual(something, null, message)` to get an "is exactly null with no weird type conversion" check.

**Pull request merged**